src/subtype.c

```diff
@@ -0,0 +1,569 @@
+    }
+
+    // handle the "diagonal dispatch" rule, which says that a type var occurring more
+    // than once, and only in covariant position, is constrained to concrete types. E.g.
+    //  ( Tuple{Int, Int}    <: Tuple{T, T} where T) but
+    // !( Tuple{Int, String} <: Tuple{T, T} where T)
+    // This is done by checking that the variable's lower bound is not an abstract type.
+    if (ans && (vb.concrete || (!vb.occurs_inv && vb.occurs_cov > 1))) {
+        if (jl_is_typevar(vb.lb)) {
+            jl_tvar_t *v = (jl_tvar_t*)vb.lb;
+            jl_varbinding_t *vlb = lookup(e, v);
+            if (vlb)
+                vlb->concrete = 1;
+            else  // TODO ???
+                ans = (v == vb.concretevar);
+        }
+        else if (!is_leaf_bound(vb.lb)) {
+            ans = 0;
+        }
+        if (ans) {
+            // if we occur as another var's lower bound, record the fact that we
+            // were concrete so that subtype can return true for that var.
+            btemp = vb.prev;
+            while (btemp != NULL) {
+                if (btemp->lb == (jl_value_t*)u->var)
+                    btemp->concretevar = u->var;
+                btemp = btemp->prev;
+            }
+        }
+    }
+
+    e->vars = vb.prev;
+    JL_GC_POP();
+    return ans;
+}
+
+static jl_value_t *unwrap_2_unionall(jl_value_t *t, jl_value_t **p1, jl_value_t **p2)
+{
+    if (jl_is_unionall(t)) {
+        *p1 = ((jl_unionall_t*)t)->var;
+        t = ((jl_unionall_t*)t)->body;
+        if (jl_is_unionall(t)) {
+            *p2 = ((jl_unionall_t*)t)->var;
+            t = ((jl_unionall_t*)t)->body;
+        }
+    }
+    return t;
+}
+
+static int subtype_tuple(jl_datatype_t *xd, jl_datatype_t *yd, jl_stenv_t *e)
+{
+    size_t lx = jl_nparams(xd), ly = jl_nparams(yd);
+    if (lx == 0 && ly == 0)
+        return 1;
+    if (ly == 0)
+        return 0;
+    size_t i=0, j=0;
+    int vx=0, vy=0;
+    while (i < lx) {
+        if (j >= ly) return 0;
+        jl_value_t *xi = jl_tparam(xd, i), *yi = jl_tparam(yd, j);
+        if (jl_is_vararg_type(xi)) vx = 1;
+        if (jl_is_vararg_type(yi)) vy = 1;
+        if (vx && !vy)
+            return 0;
+        if (!vx && vy) {
+            if (!subtype(xi, jl_unwrap_vararg(yi), e))
+                return 0;
+        }
+        else {
+            if (!subtype(xi, yi, e))
+                return 0;
+        }
+        i++;
+        if (j < ly-1 || !vy)
+            j++;
+    }
+    vy = vy || (j < ly && jl_is_vararg_type(jl_tparam(yd,j)));
+    if (vy && !vx && lx+1 >= ly) {
+        jl_tvar_t *va_p1=NULL, *va_p2=NULL;
+        jl_value_t *tail = unwrap_2_unionall(jl_tparam(yd,ly-1), &va_p1, &va_p2);
+        assert(jl_is_datatype(tail));
+        // in Tuple{...,tn} <: Tuple{...,Vararg{T,N}}, check (lx+1-ly) <: N
+        jl_value_t *N = jl_tparam1(tail);
+        // only do the check if N is free in the tuple type's last parameter
+        if (N != (jl_value_t*)va_p1 && N != (jl_value_t*)va_p2) {
```

Comment:
No, this is just handling the awkwardness of `Vararg{T}` actually being a UnionAll type. It's kind of a hand optimization of converting a non-Vararg tuple type to `Vararg{T,N}` where `N` is the length of the tail of the non-vararg type, and then recursively subtyping the two `Vararg` types.